electrumsv/wallet.py

```diff
@@ -550,21 +550,29 @@ def get_public_keys(self, address: Address):
         return self.get_pubkeys(*sequence)
 
     def add_verified_tx(self, tx_hash, height, timestamp, position, proof_position, proof_branch):
-        entry = self._datastore.tx.get_entry(tx_hash, TxFlags.StateCleared)
+        entry = self._datastore.tx.get_entry(tx_hash, TxFlags.StateCleared)  # HasHeight
+
         # Ensure we are not verifying transactions multiple times.
         if entry is None:
-            entry = self._datastore.tx.get_entry(tx_hash)
-            self.logger.debug("Attempting to clear unsettled tx %s %r",
-                tx_hash, entry)
-            return
+            # We have proof now so regardless what TxState is, we can 'upgrade' it to StateSettled.
+            # This rests on the commitment that any of the following four tx States *will*
+            # Have tx bytedata i.e. "HasByteData" flag is set.
+            possible_states = (TxFlags.StateSigned | TxFlags.StateDispatched |
+                               TxFlags.StateCleared | TxFlags.StateReceived)
+            entry = self._datastore.tx.get_entry(tx_hash, flags=possible_states)
+            self.logger.debug("Fast_tracking entry to StateSettled: %r", entry)
 
         # We only update a subset.
         flags = TxFlags.HasHeight | TxFlags.HasTimestamp | TxFlags.HasPosition
         data = TxData(height=height, timestamp=timestamp, position=position)
+
         self._datastore.tx.update([ (tx_hash, data, None, flags | TxFlags.StateSettled) ])
 
         proof = TxProof(proof_position, proof_branch)
         self._datastore.tx.update_proof(tx_hash, proof)
+        self.logger.debug("final tx_entry after updating proof: entry: %r cached entry: %r",
+                          self._datastore.tx.get_entry(tx_hash),
+                          self._datastore.tx.get_cached_entry(tx_hash))
```

Comment:
Debug logging. If this should be working correctly, this shouldn't be needed. Same for all new debug logging.